### PR TITLE
Add comments to binding gen tests for invocations on new lines

### PR DESCRIPTION
**Pull request by layomia**
Addresses https://github.com/dotnet/runtime/pull/91218#discussion_r1307801023.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs [hunk 1]
```diff
@@ -8,8 +8,14 @@ namespace Microsoft.Extensions.SourceGeneration.Configuration.Binder.Tests
...
+        /// This is a regression test for https://github.com/dotnet/runtime/issues/90851.
+        /// It asserts that the configuration binding source generator properly formats
+        /// binding invocation source locations that the generated interceptors replace.
+        /// Do not modify these tests if the change reduces coverage for this scenario.
+        /// </summary>
         [Fact]
-        public void GeneratorHandlesInvocationsOnNewline()
+        public void TestBindingInvocationsOnNewLines()
         {
             IConfiguration configuration = TestHelpers.GetConfigurationFromJsonString(@"{""Longitude"":1,""Latitude"":2}");
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs:21]: I suggest add inline comment for different cases. doing that ensure anyone touching the code will read the comment :-)

File: src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/Common/ConfigurationExtensionsTests.cs [hunk 2]
```diff
@@ -72,5 +72,52 @@ private static OptionsBuilder<FakeOptions> CreateOptionsBuilder()
...
+        /// </summary>
+        [Fact]
+        public void TestBindingInvocationsOnNewLines()
+        {
+            OptionsBuilder<FakeOptions>? optionsBuilder = CreateOptionsBuilder();
+
+            // Options builder extensions.
+
+            optionsBuilder
+                .Bind(s_emptyConfig, configureBinder: null);
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/Common/ConfigurationExtensionsTests.cs:90]: ditto, inline comments would be better.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs:21]: I don't think any individual test case has interesting details specific to its syntax. We are just exercising various legal C# invocation syntax, in the spirit of fuzz testing (which btw would be overkill IMO).

IOW any inline comments would have the same general info for every test case.

Comment (jeffhandley) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs:21]: I agree with @tarekgh's suggestion. These tests need to have comments such as, "This tests the scenario of a newline between the OptionsBuilder instance reference and `.Bind`," or, "Place a newline between the `.` and `BindConfiguration`."

As is, the comment above just says not to modify the tests, but it doesn't cite what key elements cannot be modified. Presumably, modifications to the variable names would be OK while it's the very specific newline placements that need to be retained for coverage. If the comments feel repetitive, then the comments likely have room for being more specific to the condition being tested.

**Review by jeffhandley — DISMISSED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/ConfigurationBinderTests.Generator.cs:23]: Thanks folks; added some more comments.

File: src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/SourceGenerationTests/ConfigurationExtensionsTest.Generator.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        public void TestBindingInvocationsWithIrregularCSharpSyntax()
+        {
+            // Tests binding invocation variants with irregular C# syntax, interspersed with white space.
+
+            // Options builder extensions.
+
+            OptionsBuilder<FakeOptions>? optionsBuilder = CreateOptionsBuilder();
+
+            optionsBuilder
+                .Bind(s_emptyConfig, configureBinder: null);
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/SourceGenerationTests/ConfigurationExtensionsTest.Generator.cs:30]: nit: is it worth adding similar inline comments here too?

**Review by tarekgh — APPROVED** [review]
I added an extra suggestion, LGTM otherwise. Thanks!

File: src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/SourceGenerationTests/ConfigurationExtensionsTest.Generator.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        /// It asserts that the configuration binding source generator properly formats
+        /// binding invocation source locations that the generated interceptors replace.
+        /// A location issue that's surfaced is emitting the right location of invocations
+        /// that are on a different line than the containing binder type or the static
+        /// extension binder class (e.g. ConfigurationBinder.Bind).
+        /// </summary>
+        [Fact]
+        public void TestBindingInvocationsWithIrregularCSharpSyntax()
+        {
+            // Tests binding invocation variants with irregular C# syntax, interspersed with white space.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/tests/SourceGenerationTests/ConfigurationExtensionsTest.Generator.cs:23]: I think having separate Fact method for each case will be too much especially if we'll need to add more in the future. The most important thing for me is to ensure that anyone touching this code will get a chance to see the attached comment to that line of code and understand why we wrote it this way.

**Comment by jeffhandley** [comment]
@layomia / @tarekgh -- I pushed a couple commits that expand on the comments and reorganize the tests a bit. If this looks good to you, I'll also approve and we can merge.

The spirit of what I changed:

1. Separate out the API scenarios into their own tests
2. Within each test, organize and comment on the newline scenarios covered
3. Add newline scenarios for "Newlines in every possible place"
4. In the tests with multiple asserts, do not reuse record instances across the asserts (to guard against a hidden failure).
5. Refactor the Assert method out and have it perform a single Assert on a Tuple so that both values are shown together (so that if the assert fails, the message will be clearer)

@layomia In the ConfigurationExtensions tests, there were no Asserts in the test. Are asserts needed, or does a non-throwing execution indicate success?

**Review by tarekgh — APPROVED** [review]
@jeffhandley changes LGTM. Thanks!

**Comment by layomia** [comment]
> @layomia In the ConfigurationExtensions tests, there were no Asserts in the test. Are asserts needed, or does a non-throwing execution indicate success?

Indeed non-throwing execution indicates success. The tests are really testing successful compilation. Input syntax doesn't affect the emitted binding logic, which is being tested by the other functional tests.

**Comment by layomia** [comment]
Thanks Jeff & Tarek for the detailed review; and Jeff thanks for helping to improve the comments.

**Pull request merged**